daemon/probes.go

```diff
@@ -0,0 +1,290 @@
+// number of bytes written to infoFile, number of bytes written to warningFile
+// and error.
+func probeRunLl(featureFile, infoFile io.Writer, probesDir, libIncludeDir, outDir string) (infoFileBytes int, err error) {
+	outFilename := path.Join(outDir, "raw_probe.t")
+
+	err = filepath.Walk(probesDir, func(probePath string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if strings.HasSuffix(info.Name(), ".t") {
+			outBin := path.Join(outDir, path.Base(probePath))
+			if err := copyFile(probePath, outFilename); err != nil {
+				return err
+			}
+			out, err := exec.Command(
+				"clang",
+				"-O2",
+				"-Wall",
+				fmt.Sprintf("-I%s", probesDir),
+				fmt.Sprintf("-I%s", libIncludeDir),
+				fmt.Sprintf("-I%s", outDir),
+				path.Join(probesDir, "raw_main.c"),
+				"-o",
+				outBin,
+			).CombinedOutput()
+			if err != nil {
+				return fmt.Errorf("clang failed: %s: %s", err, out)
+			}
+
+			var stdoutBuf, stderrBuf bytes.Buffer
+			cmd := exec.Command(outBin)
+			cmd.Stdout = &stdoutBuf
+			cmd.Stderr = &stderrBuf
+			if err := cmd.Run(); err != nil {
+				return err
+			}
+
+			if _, err := stdoutBuf.WriteTo(featureFile); err != nil {
+				return err
+			}
+			n, err := stderrBuf.WriteTo(infoFile)
+			if err != nil {
+				return err
+			}
+			infoFileBytes += int(n)
+		}
+		return nil
+	})
+	return
+}
+
+// writeFeatures wrtites definitions of features into the header file which is
+// going to be used by BPF programs.
+func writeFeatures(featureFile, infoFile, warningFile io.Writer, probesDir, libIncludeDir, outDir string) (infoFileBytes int, warningFileBytes int, err error) {
+	io.WriteString(featureFile, `#ifndef BPF_FEATURES_H_
+#define BPF_FEATURES_H_
+`)
+
+	var ni, nw int
+	ni, nw, err = probeKernelConfig(infoFile, warningFile)
+	if err != nil {
+		err := fmt.Errorf("kernel config probe failed: %s", err)
+		return infoFileBytes, warningFileBytes, err
+	}
+	infoFileBytes += ni
+	warningFileBytes += nw
+
+	var n int
+	n, err = probeRunLl(featureFile, infoFile, probesDir, libIncludeDir, outDir)
+	if err != nil {
+		err := fmt.Errorf("ll probe failed: %s", err)
+		return infoFileBytes, warningFileBytes, err
+	}
+	infoFileBytes += n
+
+	io.WriteString(featureFile, "#endif /* BPF_FEATURES_H_ */\n")
+
+	return
+}
+
+// runProbes runs probes for BPF features. It generates a header file for BPF
+// programs and log files.
+func runProbes(bpfDir, stateDir string) error {
+	probesDir := path.Join(bpfDir, "probes")
+	libIncludeDir := path.Join(bpfDir, "include")
+	featureFilePath := path.Join(stateDir, "globals", "bpf_features.h")
+	infoFilePath := path.Join(stateDir, "bpf_features.log")
+	warningFilePath := path.Join(stateDir, "bpf_requirements.log")
+
+	os.Remove(infoFilePath)
+	os.Remove(warningFilePath)
+
+	// File descriptors for feature, info and warning file.
+	featureFile, err := os.OpenFile(featureFilePath, os.O_RDWR|os.O_CREATE, 0644)
+	if err != nil {
+		return err
+	}
+	defer featureFile.Close()
```

Comment:
`Close()` can fail, and I see at least some documents online suggesting that it's poor style to defer file closes when you write to files, particularly in this form because the error is lost.